stl/inc/xutility

```diff
@@ -41,17 +41,47 @@ _STL_DISABLE_CLANG_WARNINGS
 #elif _USE_STD_VECTOR_ALGORITHMS
 #error Vector algorithms are not supported when building msvcp140.dll, but _USE_STD_VECTOR_ALGORITHMS is set.
 #endif // ^^^ _USE_STD_VECTOR_ALGORITHMS != 0 ^^^
-#elif (defined(_M_IX86) || defined(_M_X64)) && !defined(_M_CEE_PURE) && !defined(_M_HYBRID) && !defined(_M_ARM64EC)
-#ifndef _USE_STD_VECTOR_ALGORITHMS
-#define _USE_STD_VECTOR_ALGORITHMS 1
-#endif // !defined(_USE_STD_VECTOR_ALGORITHMS)
-#else // ^^^ arch supports vector algorithms / no support for vector algorithms vvv
+#elif defined(_M_CEE_PURE) || defined(_M_HYBRID)
 #ifndef _USE_STD_VECTOR_ALGORITHMS
 #define _USE_STD_VECTOR_ALGORITHMS 0
 #elif _USE_STD_VECTOR_ALGORITHMS
 #error Vector algorithms are not supported on this architecture, but _USE_STD_VECTOR_ALGORITHMS is set.
 #endif // ^^^ _USE_STD_VECTOR_ALGORITHMS != 0 ^^^
-#endif // ^^^ no support for vector algorithms ^^^
+#else // ^^^ vector algorithms unsupported for /clr:pure or CHPE / vector algorithms supported by default vvv
+#ifndef _USE_STD_VECTOR_ALGORITHMS
+#define _USE_STD_VECTOR_ALGORITHMS 1
+#endif // !defined(_USE_STD_VECTOR_ALGORITHMS)
+#endif // ^^^ vector algorithms supported by default ^^^
+
+// _USE_STD_VECTOR_ALGORITHMS is our documented control macro:
+// https://learn.microsoft.com/en-us/cpp/standard-library/vectorized-stl-algorithms?view=msvc-170
+// In the STL's implementation, we inspect the following architecture-specific detection macros,
+// because various algorithms are vectorized for various architectures:
+#if defined(_VECTORIZED_FOR_X64_X86)
+#error _VECTORIZED_FOR_X64_X86 is not a control macro; define _USE_STD_VECTOR_ALGORITHMS instead.
+#elif defined(_VECTORIZED_FOR_X64_X86_ARM64)
+#error _VECTORIZED_FOR_X64_X86_ARM64 is not a control macro; define _USE_STD_VECTOR_ALGORITHMS instead.
+#elif defined(_VECTORIZED_FOR_X64_X86_ARM64_ARM64EC)
+#error _VECTORIZED_FOR_X64_X86_ARM64_ARM64EC is not a control macro; define _USE_STD_VECTOR_ALGORITHMS instead.
+#elif !_USE_STD_VECTOR_ALGORITHMS
+#define _VECTORIZED_FOR_X64_X86               0
+#define _VECTORIZED_FOR_X64_X86_ARM64         0
+#define _VECTORIZED_FOR_X64_X86_ARM64_ARM64EC 0
+#elif defined(_M_ARM64EC)
+#define _VECTORIZED_FOR_X64_X86               0
+#define _VECTORIZED_FOR_X64_X86_ARM64         0
+#define _VECTORIZED_FOR_X64_X86_ARM64_ARM64EC 1
+#elif defined(_M_X64) || defined(_M_IX86)
+#define _VECTORIZED_FOR_X64_X86               1
+#define _VECTORIZED_FOR_X64_X86_ARM64         1
+#define _VECTORIZED_FOR_X64_X86_ARM64_ARM64EC 1
+#elif defined(_M_ARM64)
+#define _VECTORIZED_FOR_X64_X86               0
+#define _VECTORIZED_FOR_X64_X86_ARM64         1
+#define _VECTORIZED_FOR_X64_X86_ARM64_ARM64EC 1
+#else // ^^^ known architecture / unknown architecture vvv
+#error Unknown architecture.

```

Comment:
non-blocking nit: 'Unknown architecture' is not a super actionable error message. It would be great if there was more we could say, such as 'supported architectures can be found here: X' or 'try Y`, etc.

But I realize that's going above and beyond, and that this error is super rare. So we can dismiss. Just wanted to call this out.

Reply:
Thanks. This is a pre-existing pattern:

https://github.com/microsoft/STL/blob/1717ac0bde5412309975b1782efe2a73364cfba5/stl/inc/__msvc_sanitizer_annotate_container.hpp#L165-L167
https://github.com/microsoft/STL/blob/1717ac0bde5412309975b1782efe2a73364cfba5/stl/src/xonce2.cpp#L37-L39

I'm not concerned about the actionability of the error message - it is easy to look up the line number and figure out what happened. This should also be a "can't happen" error - we know all of the architectures we target, and unless we got confused (or we introduce a new one in the future), we will have exhaustively handled them.